### DISCUSSION TITLE
MFEM phase field

### DISCUSSION THREAD

**Discussion by starkekr** [mfem/mfem#4070] [General]
Hello,

I am very new to MFEM and am really just looking for information. If I was trying to implement a new set of physics into MFEM, say a phase field module, what all would I need to implement? To me it seems like the I would have to implement a class which derives from the NonlinearFormIntegrator class? I am very familier with the MOOSE framework code where pde's are cast into kernals and these kernels provide the residual vector and jacobian matrix for applying Newton's method to solve the discretized set of equations. I do not know which classes perform these types of operations in MFEM. I will also need to provide coupling between multiple independent variables in the problem 

Any information or pointing me to a set of classes with implement some sort of nonlinear Galerkin finite element solution scheme would be very much appreciated.

**Comment by vladotomov** [comment]
@starkekr you can start with [mfem.org/examples/](https://mfem.org/examples/) and select `Newton method` in the `Solver` drop-down list. From there you can follow links to the relevant example codes.

**Comment by starkekr** [comment]
Is there an example with a simple time dependent nonlinear galerkin problem? It seems like example 16 is close to this but not exactly what I need. How would I modify this example so that the diffusion operator is evaluated at the most recent newton step? I am asking mainly about logistics here. Does the member variable  CGSolver T_solver; need to be replaced with NewtonSolver type? In this case how does this solver get called, 
Is it through ImplicitSolve?

**Reply by vladotomov** [reply]
Did you go over example 10? That seems more relevant for what you need.

**Reply by starkekr** [reply]
I will check it out thanks!

**Reply by starkekr** [reply]
Example 10 seems to be what I need. I have one issue though, I am trying to extend this model by using BlockNonlinearForms. I am having trouble with the equivalent of line 589 which is:
 SparseMatrix *grad_H = dynamic_cast<SparseMatrix *>(&H->GetGradient(z));

Does this line work when H is a BlockNonlinearForm?

**Reply by vladotomov** [reply]
@starkekr it will be a little different, because in that case `GetGradient` will give you a `BlockOperator`. Then you can get the individual blocks with `GetBlock`, but what you do depends on how you solve the Jacobian block system. You can see example 5 for solving a block system.

**Reply by starkekr** [reply]
OK thank you. Two more questions.
1). To get the derivative of the shape functions in physical space (not some reference element) do I need to do something like the following:
                el.CalcDShape(ip, dshape);
		Mult(dshape, Tr.AdjugateJacobian(), dshapep);
where dshapep is the derivatives in physical space

2). I am confused about the weights when performing integration. Is the following a correct interpretation?
double w = Tr.Weight() * ip.weight;
The first term is the determinant of the coordinate transformation from the reference element to the physical element?
The second term is the gauss point weights? (For 2d calculations this is a product of the 1D weights at the various reference element points)

**Reply by vladotomov** [reply]
1) See `FiniteElement::CalcPhysDShape()`.
2) Correct.

**Reply by starkekr** [reply]
Thanks CalcPhysDShape worked.


A couple questions on mesh adaptivity:
1). Is there a parameter that I can set to limit the max level of refinement? If not, is there a way to see how many times an element has been refined?  I see that there is a way to limit the number of hanging nodes but this solves a different problem.
2). Does the ParaViewDataCollection class support AMR for time dependent simulations? What do I have to change from the no AMR case to make this work?

**Reply by starkekr** [reply]
Sorry 2). seems to work, but I still have a question about limiting the max level of refinement above

**Reply by starkekr** [reply]
I cant seem to get the block solver working. I have attached a minimal file here. 
[exmyNLOpMFEMBlock.zip](https://github.com/mfem/mfem/files/14357969/exmyNLOpMFEMBlock.zip)

My question is in the code surrounding in the function 

NLKernelOperator::GetGradient()

I think this method is supposed to build the member variable BlockOperator *_BlockJac and return a sparse matrix representation of it for the newton solver to use. 

My question is fairly general. Is this the correct way to use the BlockOperator class in this function when then NLKernelOperator has two member variable that are of type BlockNonlinearForm? I am unable to use the GetBlock function that you mentioned for BlockNonlinearForm as it is not a member function of that class.

**Comment by vladotomov** [comment]
> Is there a parameter that I can set to limit the max level of refinement?

I'm not sure. This can probably be controlled easily by the application.

> is there a way to see how many times an element has been refined?

See `NCMesh::GetElementDepth()`.

For questions that are unrelated to the original topic, it's better to start a new issue.
